service/src/main/java/bio/terra/externalcreds/services/FenceProviderService.java

```diff
@@ -0,0 +1,135 @@
+package bio.terra.externalcreds.services;
+
+import bio.terra.common.exception.BadRequestException;
+import bio.terra.externalcreds.ExternalCredsException;
+import bio.terra.externalcreds.auditLogging.AuditLogEvent;
+import bio.terra.externalcreds.auditLogging.AuditLogEventType;
+import bio.terra.externalcreds.auditLogging.AuditLogger;
+import bio.terra.externalcreds.config.ExternalCredsConfig;
+import bio.terra.externalcreds.generated.model.Provider;
+import bio.terra.externalcreds.models.FenceAccountKey;
+import bio.terra.externalcreds.models.LinkedAccount;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.util.HashSet;
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.security.oauth2.core.OAuth2AuthorizationException;
+import org.springframework.stereotype.Service;
+
+@Service
+@Slf4j
+public class FenceProviderService extends ProviderService {
+
+  private final BondService bondService;
+  private final FenceAccountKeyService fenceAccountKeyService;
+  private final FenceKeyRetriever fenceKeyRetriever;
+
+  public FenceProviderService(
+      ExternalCredsConfig externalCredsConfig,
+      ProviderOAuthClientCache providerOAuthClientCache,
+      ProviderTokenClientCache providerTokenClientCache,
+      OAuth2Service oAuth2Service,
+      LinkedAccountService linkedAccountService,
+      AuditLogger auditLogger,
+      ObjectMapper objectMapper,
+      BondService bondService,
+      FenceAccountKeyService fenceAccountKeyService,
+      FenceKeyRetriever fenceKeyRetriever) {
+    super(
+        externalCredsConfig,
+        providerOAuthClientCache,
+        providerTokenClientCache,
+        oAuth2Service,
+        linkedAccountService,
+        auditLogger,
+        objectMapper);
+    this.bondService = bondService;
+    this.fenceAccountKeyService = fenceAccountKeyService;
+    this.fenceKeyRetriever = fenceKeyRetriever;
+  }
+
+  public Optional<LinkedAccount> getLinkedFenceAccount(String userId, Provider provider) {
+    var ecmLinkedAccount = linkedAccountService.getLinkedAccount(userId, provider);
+    var bondLinkedAccount = bondService.getLinkedAccount(userId, provider);
+    if (ecmLinkedAccount.isPresent() && bondLinkedAccount.isPresent()) {
+      var ecmExpires = ecmLinkedAccount.get().getExpires();
+      var bondExpires = bondLinkedAccount.get().getExpires();
+      if (ecmExpires.after(bondExpires) || ecmExpires.equals(bondExpires)) {
+        // ECM is up to date
+        return ecmLinkedAccount;
+      } else {
+        // ECM is out of date. Update it with the Bond data and return the new ECM Linked Account
```

Comment:
When would the ECM linked account have an older expiration time than the one in Bond (since it would have been ported from the Bond db)?

Reply:
If, somehow, someone connected to Bond directly to renew their link. Probably won't happen, but still a possibility.

Reply:
Oh I see!